src/libsyntax/parse/parser.rs

```diff
@@ -1824,6 +1824,14 @@ impl<'a> Parser<'a> {
     fn parse_arg_general(&mut self, require_name: bool) -> PResult<'a, Arg> {
         maybe_whole!(self, NtArg, |x| x);
 
+        if let Ok(Some(_)) = self.parse_self_arg() {
+            let mut err = self.struct_span_err(self.prev_span,
+                "unexpected `self` argument in function");
+            err.span_label(self.prev_span,
+                "`self` is only valid as the first argument of a trait function");
```

Comment:
This is misleading, it also works for inherent impls.

Reply:
will fix

Reply:
Should I rename this to

> `self` is only valid as the first argument of an associated function

Reply:
That seems accurate.